File: leet_lint_code/leet_code/next_permutation.cpp
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of leet code problem 31.

#ifndef LEETLINTCODE_LINTCODE_NEXTPERMUTATION_CPP
#define LEETLINTCODE_LINTCODE_NEXTPERMUTATION_CPP

#include <vector>
#include <algorithm>

class Solution {
 public:
   void nextPermutation(vector<int> &nums) {
     std::vector<int> t;
     size_t len = nums.size();
     if (1 >= len) return;

     int j = len - 1;
     while (0 < j && nums[j] <= nums[j - 1]) -- j;

     if (0 == j) {
        std::vector<int>::iterator begin = nums.begin();
        std::reverse(begin, nums.end());
        return ;
     }

     int pivot = nums[j - 1];
     int i = len - 1;

     while (0 < i && nums[i] <= pivot) -- i;
     swap(nums[i], nums[j - 1]);
     std::vector<int>::iterator begin = nums.begin() + j;
     std::reverse(begin, nums.end());
     for (size_t i = 0; i < len; ++ i) t.push_back(nums[i]);
     nums = t;
    }
};

#endif // LEETLINTCODE_LINTCODE_NEXTPERMUTATION_CPP

File: leet_lint_code/leet_code/java/NextPermutation.java
// Copyright © 2018 Hailin Gu. All rights reserved.
// License
// Author: Hailin Gu
// This is a answer of leet code problem 31 .
// Date: 2019.8.27

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class NextPermutation {

    public void nextPermutation(int[] nums) {
        int i = nums.length - 1;
        while (i >= 1 && nums[i - 1] >= nums[i]) -- i;
        if (i == 0) {
            Arrays.sort(nums);
            return ;
        }

        List<Integer> permutationList = new ArrayList<>();
        int j = nums.length - 1;
        while (j > i && nums[j] <= nums[i - 1]) -- j;

        int p = i - 1;
        while (p < nums.length) {
            if (p != j) {
                permutationList.add(nums[p]);
            }
            ++ p;
        }

        nums[i - 1] = nums[j];
        Collections.sort(permutationList);
        p = 0;
        while (p < permutationList.size()) {
            nums[i + p] = permutationList.get(p);
            ++ p;
        }
    }
}
